File: bin/reports_to_table.rb
#!/bin/env ruby

require 'optparse'
require 'ostruct'
require 'json'
require 'csv'

### Define modules and classes here

def parse_json(filename)

    return JSON.parse(IO.readlines(filename).join)

end

### Get the script arguments and open relevant files
options = OpenStruct.new()
opts = OptionParser.new()
opts.banner = "Reads reports and makes an Excel table with one sheet per analysis rule"
opts.separator ""
opts.on("-o","--outfile", "=OUTFILE","Output file") {|argument| options.outfile = argument }
opts.on("-h","--help","Display the usage information") {
    puts opts
    exit
}

opts.parse! 

negative_result = "-"

files = Dir["*.json"]

bucket = {}

# awfully convuluted way to bin all the reports by rule and sample name
files.group_by{|f| f.split(".")[0..-3].join}.each do |group,reports|

    blast = reports.find {|r| r.include?(".blast.")}
    freebayes = reports.find { |r| r.include?(".freebayes.")}

    this_data = []

    if blast
        json = parse_json(blast)
        matches = json["matches"]

        matches.each do |match|
            rule = match["rule"]
            match["Sample"] = json["sample"]

            bucket.has_key?(rule) ? bucket[rule] << match : bucket[rule] = [ match ]
        end

    end

    if freebayes
        json = parse_json(freebayes)
        matches = json["matches"]
        matches.each do |match|
            rule = match["rule"]
            match["Sample"] = json["sample"]
            bucket.has_key?(rule) ? bucket[rule] << match : bucket[rule] = [ match ]
        end
    end

end


# a bucket is a rule with all the matching reports, i.e. one page
bucket.each do |rule,reports|

    csv_list = []

    this_row = []
    row = 0
    col = 0

    # The table header
    [ "Probe", "Vsearch/Blast", "Bwa2/Freebayes"].each do |e|
        this_row << e
        col += 1
    end

    csv_list << this_row
    # Each sample with all its reports (max 2)
    reports.group_by{|r| r["Sample"]}.sort.each do |sample,data|

        this_row = []

        row += 1
        col = 0
        
        this_row << sample

        col += 1
        
        blast = data.find{ |d| d["toolchain"].include?("vsearch")}
        freebayes = data.find{ |d| d["toolchain"].include?("bwa2")}

        blast ? b = blast["perc_gmo"] : b = negative_result
        # Here we remove noisy results below 1%
        if b.is_a?(Float)
            b = negative_result if b.to_f < 1.0
        end

        if freebayes
            freebayes.has_key?("perc_gmo") ? f = freebayes["perc_gmo"] : f = negative_result
        else
            f = "N/A"
        end

        this_row << b
        this_row << f

        csv_list << this_row

    end

    header = [ "# id: 'gmo_check_result'",
        "# section_name: '#{rule}'",
        "# description: 'GMO Nachweis für #{rule} (Anteil in %).'",
        "# format: 'tsv'",
        "# plot_type: 'table'",
        "# pconfig:",
        "#    id: 'custom_bargraph_w_header'",
        "#    ylab: 'Anteil GMO'" ]


    file_name = rule.gsub(" ","_").downcase
    file = File.new(file_name +"_mqc.tsv","w+")

    file.puts header.join("\n")

    csv_list.each do |entry|

        file.puts entry.join("\t")

    end

    file.close
    
end

File: bin/reports_to_table.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


import argparse
import glob
import json

parser = argparse.ArgumentParser(description="Script options")
parser.add_argument("--output")
args = parser.parse_args()


def main(output):

    # Read all the JSON reports
    reports = sorted(glob.glob("*.json"))
    bucket = {}

    # Pre-configured list of toolchains to report
    toolchains = [ "vsearch", "bwa2"]

    for report in reports:

        # Parse the JSON file
        with open(report) as f:
            refs = json.load(f)

        sample = refs["sample"]
        matches = refs["matches"]
        
        for match in matches:
            rule = match["rule"]
            match["Sample"] = sample
            if rule in bucket:
                bucket[rule].append(match)
            else:
                bucket[rule] = [ match ]

    # Results by rule - creates on result file each
    for rule, matches in bucket.items():

        header = [ "# id: 'gmo_check_result'",
            f"# section_name: '{rule}'",
            f"# description: 'GMO Nachweis für {rule} (Anteil in %).'",
            "# format: 'tsv'",
            "# plot_type: 'table'",
            "# pconfig:",
            "#    id: 'custom_bargraph_w_header'",
            "#    ylab: 'Anteil GMO'" ]
        
        csv_list = []

        this_row = []
        row = 0
        col = 0

        # The table header
        for entry in [ "Probe", "Vsearch/Blast", "Bwa2/Freebayes"]:
            this_row.append(entry)
            col += 1

        csv_list.append(this_row)

        grouped_reports = {}
        for match in matches:
            sample = match["Sample"]
            if sample in grouped_reports:
                grouped_reports[sample].append(match)
            else:
                grouped_reports[sample] = [ match ]

    
        for sample,reports in grouped_reports.items():
            this_row = []
            row += 1
            col = 0

            this_row.append(sample)

            col += 1

            for toolchain in toolchains:
                # Check if we have an entry for this tool chain
                data = next((item for item in reports if item["toolchain"] == toolchain ),None)
                if data:
                    this_row.append(str(data["perc_gmo"]))
                else:
                    this_row.append("-")
            
            csv_list.append(this_row)

        rule_name = rule.lower().replace(" ", "_")
        this_result = f"{rule_name}_mqc.tsv"
        with open(this_result, "w") as fo:
            fo.write("\n".join(header)+ "\n")
            for row in csv_list:
                entry = "\t".join(row)
                fo.write(entry+ "\n")
        

if __name__ == '__main__':
    main(args.output)
